src/v/container/include/container/chunked_vector.h

```diff
@@ -0,0 +1,303 @@
+        using difference_type = std::ptrdiff_t;
+        using pointer = value_type*;
+        using reference = value_type&;
+
+        iter() = default;
+
+        reference operator*() const { return _vec->operator[](_index); }
+        pointer operator->() const { return &_vec->operator[](_index); }
+
+        iter& operator+=(ssize_t n) {
+            _index += n;
+            return *this;
+        }
+
+        iter& operator-=(ssize_t n) {
+            _index -= n;
+            return *this;
+        }
+
+        iter& operator++() {
+            ++_index;
+            return *this;
+        }
+
+        iter& operator--() {
+            --_index;
+            return *this;
+        }
+
+        iter operator++(int) {
+            auto tmp = *this;
+            ++*this;
+            return tmp;
+        }
+
+        iter operator--(int) {
+            auto tmp = *this;
+            --*this;
+            return tmp;
+        }
+
+        iter operator+(difference_type offset) { return iter{*this} += offset; }
+        iter operator-(difference_type offset) { return iter{*this} -= offset; }
+
+        bool operator==(const iter&) const = default;
+        auto operator<=>(const iter&) const = default;
+
+        friend ssize_t operator-(const iter& a, const iter& b) {
+            return a._index - b._index;
+        }
+
+    private:
+        friend class chunked_vector;
+        using vec_type = std::conditional_t<Const, const this_type, this_type>;
+
+        iter(vec_type* vec, size_t index)
+          : _index(index)
+          , _vec(vec) {}
+
+        size_t _index{};
+        vec_type* _vec{};
+    };
+
+    using const_iterator = iter<true>;
+    using iterator = iter<false>;
+
+    iterator begin() { return iterator(this, 0); }
+    iterator end() { return iterator(this, _size); }
+
+    const_iterator begin() const { return const_iterator(this, 0); }
+    const_iterator end() const { return const_iterator(this, _size); }
+
+    const_iterator cbegin() const { return const_iterator(this, 0); }
+    const_iterator cend() const { return const_iterator(this, _size); }
+
+    friend std::ostream& operator<<(std::ostream& os, const chunked_vector& v) {
+        os << "[";
+        for (auto& e : v) {
+            os << e << ",";
+        }
+        os << "]";
+        return os;
+    }
+
+private:
+    struct chunk {
+        // We don't need the capacity pointer like std::vector...
+        T* start;
+        T* end;
+    };
+
+    void maybe_add_capacity() {
+        if (_chunks.back().size() >= elements_per_chunk) {
+            std::vector<T> frag;
+            frag.reserve(elements_per_chunk);
+            _chunks.push_back(std::move(frag));
+        }
+    }
+
+private:
+    friend class chunked_vector_validator;
+    chunked_vector(const chunked_vector&) noexcept = default;
+
+    size_t _size{0};
+    // Invariant: Always keep around a single vector.
+    absl::InlinedVector<std::vector<T>, 1> _chunks = {{}};
```

Comment:
Interesting to think how this compares to the original approach with `std::vector`.

Pros to this approach: 
 - Avoids one allocation for some sizes of vector, especially all 1-chunk vectors, though not empty vectors
 - When accessing a 1-chunk vector, one level of pointer indirection is avoided 
 
Cons:
  - sizeof(chunked_vector) is presumably larger since I assume there is additional metadata in addition to the inlined element
  - extra branch for all access, though well predicted
  - for chunks > 2, indexing access is presumably slower as it needs to do the indirection anyway, plus the check
  
Much of extra accessing cost could be removed when iterating, however, with a differnet iterator implementation.


I feel like it will pessimise access

Reply:
> sizeof(chunked_vector) is presumably larger since I assume there is additional metadata in addition to the inlined element

It's 8 bytes larger FWIW

I'll be removing this in future pushes. It's still an interesting optimization, but probably not warranted with out a deeper investigation on our vectors than I'm willing to do.